native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -189,71 +213,153 @@ private static Type getFieldType(String fieldName, Type type) {
+                        !isBelongingToElement(parentAttributeMap, entry.getKey(), value)) {
                     String key = attributePrefix + getKey(entry, nsPrefixMap, preserveNamespaces);
-                    putAsFieldTypes(mapData, key, value.getValue(), type);
+                    putAsFieldTypes(mapData, key, value.getValue(), getFieldType(key, type));
                 }
             }
         } else {
             for (Map.Entry<BString, BString> entry : attributeMap.entrySet()) {
                 String key = getKey(entry, nsPrefixMap, preserveNamespaces);
                 if (key != null) {
-                    putAsFieldTypes(mapData, attributePrefix + key, entry.getValue().getValue(), type);
+                    key = attributePrefix + key;
+                    putAsFieldTypes(mapData, key, entry.getValue().getValue(), getFieldType(key, type));
                 }
             }
         }
     }
 
     private static boolean isBelongingToElement(BMap<BString, BString> parentAttributeMap, BString key,
                                                        BString value) {
-        return !(parentAttributeMap.containsKey(key) &&
-                parentAttributeMap.get(key).getValue().equals(value.getValue()));
+        return parentAttributeMap.containsKey(key) && parentAttributeMap.get(key).getValue().equals(value.getValue());
     }
 
-    private static void putAsFieldTypes(BMap<BString, Object> map, String key, String value, Type type) {
+    private static void putAsFieldTypes(BMap<BString, Object> map, String key, String value, Type type)
+            throws Exception {
         if (type != null) {
-            String valueType;
             if (type instanceof ArrayType) {
                 Type fieldType = ((ArrayType) type).getElementType();
                 if (fieldType instanceof RecordType) {
                     if (((RecordType) fieldType).getFields().get(key) != null) {
-                        type = ((RecordType) fieldType).getFields().get(key).getFieldType();
-                        valueType = type.getName();
-                    } else {
-                        valueType = fieldType.getName();
+                        type = ((RecordType) type).getFields().get(key).getFieldType();
                     }
-                } else {
-                    valueType = ((ArrayType) type).getElementType().getName();
+                }
+            }
+            if (type instanceof UnionType) {
+                UnionType bUnionType = (UnionType) type;
+                boolean isSuccessfullyCast = false;
+                for (Type memberType : bUnionType.getMemberTypes()) {
+                    try {
+                        convertToRecordType(map, memberType, key, value);
+                        isSuccessfullyCast = true;
+                        break;
+                    } catch (Exception e) {
+                        // Ignored
```

Comment:
Shouldn't we handle the exception?

Reply:
No, Here, we are converting the union-type field. So, if failed, we have to try with the next type.

Reply:
Ack.

Reply:
What if the exception is thrown for some other reason than type mismatch?

Reply:
Fixed in [297782f](https://github.com/ballerina-platform/module-ballerina-xmldata/pull/430/commits/297782fffdb16ae4ee5ace88bd49df518687a762)